src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,412 @@
+                        mod.getRight(), op_pin_state.op_pin_state_pinned);
+            }
+            case iro_Call -> {
+                Call call = (Call) replacement;
+                Node[] preds = Util.iterableToArray(call.getPreds());
+                int newSize = preds.length - 2;
+                Node[] ins = new Node[newSize];
+                if (preds.length - 2 >= 0) {
+                    System.arraycopy(preds, 2, ins, 0, preds.length - 2);
+                }
+                yield g.newCall(call.getBlock(), call.getMem(), call.getPtr(), ins, call.getType());
+            }
+            case iro_Load -> {
+                Load load = (Load) replacement;
+                yield g.newLoad(replacement.getBlock(), load.getMem(), load.getPtr(),
+                        load.getLoadMode(), load.getType(), ir_cons_flags.cons_none);
+            }
+            case iro_Store -> {
+                Store store = (Store) replacement;
+                yield g.newStore(store.getBlock(), store.getMem(), store.getPtr(),
+                        store.getValue(), store.getType(), ir_cons_flags.cons_none);
+            }
+            default -> null;
+        };
+
+
+        if (newNode != null) {
+            Pointer newPointer = newNode.getBlock().ptr;
+            Pointer origPointer = orig.getBlock().ptr;
+            Pointer replacementPointer = replacement.getBlock().ptr;
+
+            boolean dominatesOld = dominates(newPointer, origPointer);
+            boolean dominatesNew = dominates(newPointer, replacementPointer);
+
+            if (!dominatesOld || !dominatesNew) {
+                // can not replace node if the new node isnt dominating both former nodes.
+                return false;
+            }
+            // new node has to dominate all uses
+            boolean usesDominated = dominatesUses(newNode, orig);
+            boolean usesDominatedRep = dominatesUses(newNode, replacement);
+
+            if (usesDominated && usesDominatedRep) {
+                replacementMap.put(replacement, newNode);
+                replacementMap.put(orig, newNode);
+                Graph.exchange(replacement, newNode);
+                Graph.exchange(orig, newNode);
+                return true;
+            } else {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    private boolean dominates(Pointer source, Pointer target) {
+        return binding_irdom.block_dominates(source, target) != 0;
+    }
+
+    private boolean dominatesUses(Node newNode, Node orig) {
+        for (BackEdges.Edge edge : BackEdges.getOuts(orig)) {
+            Node n = edge.node;
+            if (!dominates(newNode.getBlock().ptr, n.getBlock().ptr)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static Node findReplacement(Node node, Map<Node, Node> replacements) {
+        for (int i = 0; i < MAX_INDIRECTIONS; ++i) {
+            node = replacements.get(node);
```

Comment:
`replacements` does always contain the node as key, right?

Reply:
In theory, yes `node` _should_ never be `null` ...

Reply:
And if I loop for a couple of times, replacing `node` with `replacements.get(node)`, I will never find a `node != null` with `replacements.containsKey(node) == false`?

Reply:
Actually, this function is not correct as it is right now (see the newest commit). I'm really not sure what the check for `iro_Deleted` is supposed to do.